### PR TITLE
feat: Migrate backend hosting from Azure to Cloudflare

### PR DESCRIPTION
**Pull request by DaveSkender**
Consolidates hosting onto Cloudflare. The frontend already deployed to Cloudflare Pages; the
backend was the remaining always-on cost — an App Service web app, an Azure Functions app, Blob
Storage, Key Vault, and Application Insights, all mostly idle for a demo whose data changes once
per trading day.

| Before (Azure) | After (Cloudflare) |
| :--- | :--- |
| App Service running `WebApi` | Cloudflare Container behind a Worker, sleeps when idle |
| Azure Functions timer `UpdateQuotes` | Cron trigger on the same Worker, in TypeScript |
| Blob Storage `chart-demo` | R2, read via the Worker's outbound binding |
| Key Vault / app settings | `wrangler secret` |
| Application Insights | Workers Observability |

Estimated **~$6-7/month** all-in (Workers Paid's $5 floor plus container usage). The Worker
response cache is what keeps the container asleep, so it's part of the cost story rather than an
optimization.

## What stayed the same

All ~92 indicator endpoints, the output cache, `CacheSettings`, response compression, and the
`QuoteBackup` failover path are untouched. `FacioQuo.Stock.Indicators` remains the point of the
demo, and every call into it is unchanged C#.

## What changed

**New `server/edge` workspace** — one Worker with two handlers:

- `fetch` — answers CORS preflight, serves cached responses, forwards misses to the container.
- `scheduled` — pulls Alpaca daily bars into R2. Port of `UpdateQuotes.cs`. Never touches the
  container, so refreshing quotes doesn't wake (or bill) it.

**Storage seam** — `IStorage`/`BlobClient` became `IQuoteStore` over HTTP. The container holds no
storage credentials: it GETs `http://quotes.r2/QQQ-DAILY.json` and the Worker's `outboundByHost`
handler resolves that through its R2 binding inside the Workers runtime. `enableInternet` is off.
Dropped `Azure.Storage.Blobs`, `Microsoft.Extensions.Azure`, `Azure.Identity`, and
`Azure.Security.KeyVault.Secrets`.

**Two container-hosting fixes** that would otherwise have failed at runtime rather than at build:

- Removed `UseHttpsRedirection`/`UseHsts`. TLS terminates at the edge and the Worker-to-container
  hop is plain HTTP, so these bounced every proxied request to an unreachable `https://localhost`.
- Added `Api:PublicBaseUrl`. `/indicators` built absolute URLs from `Request.Host`, which behind
  the Worker is `localhost:8080`.

**Cross-origin cache design** — entries are keyed by URL alone and stored with *all* CORS headers
stripped; the correct `Access-Control-Allow-Origin` is written per request on the way out. This
replaces the `Vary: Origin` mechanism from #517 while letting one cache entry serve every allowed
origin. Responses carry `x-edge-cache: HIT|MISS`.

**Dataset contract pinned across languages** — the Worker must emit PascalCase, because
`QuoteService` deserializes with default, case-sensitive `JsonSerializerOptions`; camelCase would
silently yield default-valued bars rather than an error. `server/quote-dataset.contract.json` is
asserted from both `quotes.spec.ts` and `QuoteDatasetContractTests.cs`.

**Deleted `server/Functions/`**, and with it Azurite and Azure Functions Core Tools as
prerequisites. A fresh clone now renders charts with `dotnet run` alone — no emulator, no cloud
credentials. Cron moved from `0 */1 08-18 * * 1-5` (EST-dependent) to `*/5 13-23 * * 1-5` (UTC;
Worker crons are UTC-only). Daily bars don't justify per-minute polling.

## Toolchain changes that rode along

The working tree already had an in-progress TypeScript 5.9.3 → 7.0.2 upgrade, entangled with this
change in `package.json` and the lockfile. TS 7 broke both `pnpm run lint` (typescript-eslint 8.64)
and `pnpm run build` (rollup-plugin-dts). Per direction, this pins **TypeScript 6.0.3** instead,
which restores both. The `packageManager` pin also moves off `pnpm@11.13.0`, which pnpm itself now
refuses to install ("a broken release... shipped without a binary"). This blocked every pnpm
command locally and would have blocked CI too.

(`libs/indy-charts/tsconfig.json` briefly carried an `ignoreDeprecations: "6.0"` workaround for a
tsup/rollup-plugin-dts-injected `baseUrl` option that TS 6 rejects; removed during review — the
current tsdown/rolldown build no longer triggers it.)

## Verification

All green: `pnpm run lint`, `pnpm run format:check`, `pnpm run build`, `dotnet build -warnAsError`
(0 warnings), `dotnet format --verify-no-changes`, Roslynator (0 diagnostics), markdownlint,
stylelint. Tests: 65 backend, 26 edge worker, 77 web, 237 libs.

Verified against the real container image (`wrangler deploy --dry-run` builds it, then `docker run`):

- plain HTTP `200` on `/` — no redirect loop;
- `Cache-Control: public, max-age=900` and `Vary: Origin` present, which is exactly what the
  Worker's `isCacheable()` matches and `stripCors()` removes;
- with no reachable quote host, `/quotes` serves the bundled 2018 backup dataset;
- with `Quotes__BaseUrl` pointed at a stand-in host, `/quotes` serves the live dataset — the full
  R2 read path, including PascalCase parsing;
- with `Api__PublicBaseUrl` set, `/indicators` emits 92 listings with correct absolute URLs;
- `ETag: YYYY.MM.DD` marker present for CI replacement.

**Unit-tested since, still not live-verified end-to-end:** `getContainer().fetch()` in the Worker's
`fetch` handler and the `outboundByHost` R2 handler now have dedicated unit coverage
(`index.spec.ts`, `container.spec.ts`, added during review) with the container hop mocked — OPTIONS
preflight, method rejection, GET/HEAD cache hits (including a fixed HEAD-cache-hit body leak),
cache-miss forwarding (success and rejection → 502), and the R2 key derivation / 404 / 200 paths
are all asserted. What's still unverified is the real Worker ↔ container round-trip under
`wrangler dev`: wrangler refuses local container development on Windows (*"you should use WSL
instead"*), and this WSL distro has no Node toolchain. CI now also builds the real
`server/Dockerfile` image (no push, `validate-container` job) on every PR that touches it, so a
broken image build is caught before `deploy-api`'s first real build on main. The container-only
workflow that reproduces the manual verification above is documented in
[server/edge/README.md](server/edge/README.md) for anyone hitting the same limitation.

Because a blocked outbound request is indistinguishable from an unpublished dataset — the API
fails over to backup data silently by design — `enableInternet` is called out as the first thing
to check if `/quotes` returns 2018 data after deploy.

## Deploying this

Full checklist in [server/edge/README.md](server/edge/README.md#deployment). The essentials:

1. Workers Paid plan; `wrangler r2 bucket create stock-charts-quotes`; `wrangler secret put`
   `ALPACA_KEY` and `ALPACA_SECRET`.
2. `CLOUDFLARE_API_TOKEN` now needs **Workers Scripts + Containers + R2** edit scope, broader than
   the Pages-only scope it had.
3. **The production API hostname is a placeholder.** `https://stock-charts-api.stockindicators.workers.dev`
   appears in `web/src/config/env.ts`, `tests/vitepress/.vitepress/theme/index.ts`,
   `tests/playwright/vitepress.spec.ts`, `server/edge/wrangler.jsonc`, and the deploy workflow. The
   account subdomain is almost certainly different. CI reads the `INDICATORS_API_BASE` repository
   variable, which overrides the workflow default without a code change.
4. `PUBLIC_BASE_URL` in `wrangler.jsonc`'s `vars` is set to that same placeholder hostname — update
   it alongside #3 once the real subdomain is known, and review `ALLOWED_ORIGINS` (seeded from
   `appsettings.json` plus the docs site; supports `*.` wildcards for previews).
5. Retire the `PUBLISH_PROFILE_FNS` and `PUBLISH_PROFILE_API` secrets.
6. Keep the Azure API alive ~1 week after cutover — already-loaded browser bundles still point at
   it — then delete the App Service, Functions app, Storage account, Key Vault, and App Insights.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -133,6 +133,38 @@ jobs:
...
+      - name: Setup Node.js
+        uses: actions/setup-node@v6
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+          cache: pnpm
+          cache-dependency-path: pnpm-lock.yaml
+
+      - name: Install dependencies
+        run: pnpm install --frozen-lockfile
+
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply .github/workflows/ci.yml:168]: Fixed in 1478a08: `test-edge` now has `packages: read` and installs with `NODE_AUTH_TOKEN`, matching `test-libraries`/`test-web`.

File: .github/workflows/deploy-website.yml [hunk 2]
```diff
@@ -177,6 +145,60 @@ jobs:
...
+  # Builds the container image from server/Dockerfile and deploys it together
+  # with the Worker that fronts it and refreshes quote datasets.
+  deploy-api:
+    name: Deploy API
+    runs-on: ubuntu-latest
+    if: github.ref == 'refs/heads/main'
+    needs:
+      - versioning
+      - build-backend
+
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply .github/workflows/deploy-website.yml:172]: Fixed in 1478a08: `deploy` now needs `deploy-api` and its `if:` requires `deploy-api.result` to be `success` or `skipped` (skipped on preview builds, required to succeed on main).

File: .vscode/launch.json [hunk 3]
```diff
@@ -4,10 +4,26 @@
...
+      "program": "${workspaceFolder}/server/WebApi/bin/Debug/net10.0/WebApi.dll",
+      "cwd": "${workspaceFolder}/server/WebApi",
+      "env": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      },
+      "serverReadyAction": {
+        "action": "openExternally",
+        "pattern": "\\bNow listening on:\\s+(https?://\\S+)"
+      }
+    },
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply .vscode/launch.json:23]: Fixed in 1478a08: added explicit `ASPNETCORE_URLS": "https://localhost:5001;http://localhost:5000"` to the "Run: Web API" launch config.

File: .vscode/tasks.json [hunk 4]
```diff
@@ -437,56 +437,32 @@
...
       "label": "Run: Backend services",
-      "detail": "Start all Backend .NET services dev server",
-      "dependsOn": [
-        "Run: Azure Storage",
-        "Run: Azure Functions",
-        "Run: Web API"
-      ],
+      "detail": "Start the Web API dev server",
+      "dependsOn": ["Run: Web API"],
       "dependsOrder": "sequence"
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply .vscode/tasks.json:475]: Addressed differently in 1478a08: rather than folding the edge Worker into "Run: Full development stack" (which would require Docker for the default dev loop), I clarified the task's `detail` text to state it starts the Web API + website directly, without the edge Worker, and points at "Run: Edge worker" for the Cloudflare-fronted stack. This matches what README.md and docs/CONTRIBUTING.md already document this task as doing.

File: libs/indy-charts/tsconfig.json [hunk 5]
```diff
@@ -18,6 +18,11 @@
     "noFallthroughCasesInSwitch": true,
     "resolveJsonModule": true,
     "allowSyntheticDefaultImports": true,
+    // tsup's declaration build hardcodes `baseUrl: compilerOptions.baseUrl || "."`
+    // when it hands options to rollup-plugin-dts, and TypeScript 6 errors on the
+    // deprecated option (TS5101). Nothing here sets baseUrl, so this only
+    // silences tsup's injected value. Remove once tsup drops it.
+    "ignoreDeprecations": "6.0",
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply libs/indy-charts/tsconfig.json]: Fixed in 1478a08: removed `ignoreDeprecations: "6.0"`. Verified `pnpm --filter @facioquo/indy-charts run build` (tsdown) and `typecheck` both succeed without it — the current tsdown/rolldown toolchain no longer injects the deprecated `baseUrl` option, so the suppression was stale.

File: server/edge/wrangler.jsonc [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  },
+
+  "vars": {
+    // Origins allowed to call the API cross-site. The Worker answers preflight
+    // and rewrites Access-Control-* itself, so this is the production source of
+    // truth; server/WebApi/appsettings.json only covers direct/local hosting.
+    "ALLOWED_ORIGINS": "https://charts.stockindicators.dev,https://dotnet.stockindicators.dev,https://stock-charts.pages.dev,https://stock-charts-vitepress.pages.dev",
+    "QUOTE_SYMBOLS": "SPY,QQQ",
+    "QUOTE_HISTORY_DAYS": "800"
+  }
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/edge/wrangler.jsonc:92]: Fixed in 1478a08: added `PUBLIC_BASE_URL` to `wrangler.jsonc`'s `vars`, set to the Worker's own `workers.dev` origin.

File: server/WebApi.Tests/Services/QuoteServiceCacheTests.cs [hunk 7]
```diff
@@ -66,6 +74,8 @@ public async Task Get_DifferentSymbols_AreCachedIndependently()
         await service.Get("SPY", CancellationToken.None);
 
         // Assert — two distinct symbols means two storage reads, not four.
-        storageMock.Verify(s => s.GetBlobClient(It.IsAny<string>()), Times.Exactly(2));
+        storeMock.Verify(
+            s => s.OpenQuotesAsync(It.IsAny<string>(), It.IsAny<CancellationToken>()),
+            Times.Exactly(2));
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/WebApi.Tests/Services/QuoteServiceCacheTests.cs]: Fixed in 1478a08: now asserts `OpenQuotesAsync("QQQ", ...)` and `OpenQuotesAsync("SPY", ...)` are each called exactly once, instead of just checking the total call count.

File: server/WebApi/Program.cs [hunk 8]
```diff
@@ -75,16 +76,30 @@
...
+// no storage credentials are needed here. Leaving it unset (the norm for local
+// development) makes QuoteService serve its bundled backup dataset.
+string? quotesBaseUrl = configuration.GetValue<string>("Quotes:BaseUrl");
+
+services.AddHttpClient(HttpQuoteStore.HttpClientName, client => {
+    if (!string.IsNullOrWhiteSpace(quotesBaseUrl))
+    {
+        // Trailing slash is required: relative object names resolve against it.
+        client.BaseAddress = new Uri(quotesBaseUrl.TrimEnd('/') + '/');
+    }
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/WebApi/Program.cs:108]: Fixed in 1478a08: `Quotes:BaseUrl` is validated with `Uri.TryCreate` (absolute, http/https only) before being assigned to `BaseAddress`; an invalid value now throws with a message naming `Quotes:BaseUrl`.

File: web/src/pages/ChartPage.tsx [hunk 9]
```diff
@@ -71,12 +71,11 @@ export function ChartPage(): React.JSX.Element {
...
                 </p>
                 <ul className="api-error-list">
-                  <li>Azure Functions (port 7071)</li>
                   <li>Web API (port 5001)</li>
-                  <li>Azurite Storage Emulator</li>
                 </ul>
                 <p className="api-error-help">
-                  Run <code>pnpm dev</code> for the frontend and start the backend services.
+                  Run <code>pnpm dev</code> for the frontend, and{" "}
+                  <code>cd server/WebApi &amp;&amp; dotnet run</code> for the backend.
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply web/src/pages/ChartPage.tsx]: Fixed in 1478a08: now reads `pnpm start`.

File: server/edge/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+ * next client. Cloudflare compresses on the way out to the browser.
+ */
+function upstreamRequest(request: Request): Request {
+  const forwarded = new Request(request);
+  forwarded.headers.delete("accept-encoding");
+  return forwarded;
+}
+
+export default {
+  async fetch(request: Request, env: Env, ctx: ExecutionContext): Promise<Response> {
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/edge/src/index.ts:50]: Fixed in 1478a08: added `server/edge/src/index.spec.ts`, covering OPTIONS preflight, method rejection, GET cache-hit (CORS re-applied + CACHE_STATUS header), HEAD cache-hit, cache-miss forwarding to a mocked container (both success and rejected-promise -> 502), the GET-only cache-write guard, and the non-cacheable-response case. `@cloudflare/containers` is mocked via `vi.mock` so no real container or `wrangler dev` is needed; `tsc --noEmit` and `vitest run` both pass.

File: server/edge/src/container.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/**
+ * Translates the container's plain HTTP GET of `http://quotes.r2/QQQ-DAILY.json`
+ * into an R2 binding call. This is how the container reads storage without ever
+ * holding credentials — the handler runs in the Workers runtime, not the
+ * container sandbox.
+ */
+ApiContainer.outboundByHost = {
+  [QUOTES_HOST]: async (request: Request, env: Env): Promise<Response> => {
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/edge/src/container.ts:55]: Fixed in 1478a08: added `server/edge/src/container.spec.ts`, covering `ApiContainer.outboundByHost[QUOTES_HOST]`: R2 key derivation (leading slash stripped), 404 when `env.QUOTES.get()` resolves null, and 200 + `content-type: application/json` when it resolves an object. `@cloudflare/containers`'s `Container` base class is stubbed via `vi.mock` since the real package imports the Workers-only `cloudflare:workers` module.

File: server/edge/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    // stored with all CORS headers stripped, and the correct
+    // Access-Control-Allow-Origin is written per request on the way out, so one
+    // entry serves every allowed origin without cross-origin poisoning.
+    const cache = caches.default;
+    const cacheKey = new Request(request.url, { method: "GET" });
+
+    const hit = await cache.match(cacheKey);
+
+    if (hit !== undefined) {
+      const response = new Response(hit.body, hit);
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/edge/src/index.ts]: Fixed in 1478a08: cache-hit responses now omit the body when `request.method === "HEAD"`, matching the GET/HEAD distinction already made on the cache-write path. Covered by the new `index.spec.ts`.

File: server/edge/src/quotes.ts [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        return;
+      }
+
+      const key = datasetKey(symbol);
+      await env.QUOTES.put(key, JSON.stringify(toStoredBars(bars)), {
+        httpMetadata: { contentType: "application/json" }
+      });
+      console.log(`Updated dataset: ${key} (${bars.length} bars)`);
+    })
+  );
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/edge/src/quotes.ts]: Fixed in 1478a08: `refreshQuotes` now uses `Promise.allSettled`, logs each rejected symbol individually with its reason, and only throws (naming the failed symbols) after every write has been attempted — so one symbol's R2 failure no longer blocks the others from publishing.

File: web/src/config/env.ts [hunk 14]
```diff
@@ -11,9 +11,14 @@ export interface EnvConfig {
...
 
-const defaultApi = import.meta.env.PROD
-  ? "https://stock-charts-api.azurewebsites.net"
-  : "https://localhost:5001";
+/**
+ * Production API origin: the Cloudflare Worker that fronts the indicator API
+ * container (see `server/edge`). Keep in sync with `PROD_API_URL` in
+ * `tests/vitepress/.vitepress/theme/index.ts`.
+ */
+const PROD_API_URL = "https://stock-charts-api.stockindicators.workers.dev";
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply web/src/config/env.ts]: Left as-is after review: the duplication is intentional, not an oversight. `tests/vitepress/.vitepress/theme/index.ts` has an explicit comment that production builds must ignore `VITE_API_URL` and always use the hardcoded `PROD_API_URL`, specifically as a guard against a misconfigured env var in the Cloudflare Pages dashboard silently repointing the site at the wrong API. Centralizing the constant would either have to preserve that hardcoding (no real dedup benefit — still two literals, just referenced from a shared file) or drop the guard (behavior change beyond what this review asked for). Both files already cross-reference each other ("Keep in sync with PROD_API_URL in ...") so drift is caught at review time. Happy to revisit if you'd rather see a shared constants file regardless.

File: server/edge/package.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    "typecheck": "tsc --noEmit",
+    "build": "tsc --noEmit",
+    "test": "vitest run",
+    "test:watch": "vitest",
+    "clean": "rm -rf .wrangler node_modules/.vite",
+    "format": "prettier --write \"src/**/*.ts\" wrangler.jsonc",
+    "format:check": "prettier --check \"src/**/*.ts\" wrangler.jsonc"
+  },
+  "devDependencies": {
+    "@cloudflare/containers": "^0.3.7",
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/edge/package.json]: Fixed in 1478a08: moved `@cloudflare/containers` to `dependencies`; lockfile updated via `pnpm install`.

File: server/WebApi.Tests/Services/HttpQuoteStoreTests.cs [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    /// <summary>Records the request it received and replays a canned response.</summary>
+    private sealed class StubHandler(HttpStatusCode status, string? body = null) : HttpMessageHandler
+    {
+        public Uri? RequestedUri { get; private set; }
+
+        protected override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/WebApi.Tests/Services/HttpQuoteStoreTests.cs:29]: Fixed in 1478a08: added `OpenQuotesAsync_WithCancelledToken_Throws`, which pre-cancels the token and asserts `OperationCanceledException` (via `Assert.ThrowsAnyAsync`) is thrown, exercising `StubHandler`'s `ThrowIfCancellationRequested()`.

File: .github/workflows/deploy-website.yml [hunk 17]
```diff
@@ -177,6 +145,60 @@ jobs:
           path: web/dist/app
           include-hidden-files: true
 
+  # Builds the container image from server/Dockerfile and deploys it together
+  # with the Worker that fronts it and refreshes quote datasets.
+  deploy-api:
+    name: Deploy API
+    runs-on: ubuntu-latest
+    if: github.ref == 'refs/heads/main'
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply .github/workflows/deploy-website.yml:168]: Fixed in 1478a08: added a `validate-container` job that runs `docker build --platform linux/amd64 -f server/Dockerfile server/` (no push) on every trigger of this workflow, so a broken Dockerfile is caught on the PR that changes it rather than at `deploy-api`'s first real build on main.

File: server/WebApi/Endpoints.cs [hunk 18]
```diff
@@ -35,7 +39,13 @@ public IActionResult GetIndicatorCatalog()
         Response.Headers.ETag = "YYYY.MM.DD"; // replaced in build deployment
         Response.Headers.LastModified = DateTime.UtcNow.ToString("R");
 
-        return Ok(Metadata.IndicatorListing($"{Request.Scheme}://{Request.Host}"));
+        // Behind the edge Worker this container only sees its own internal
+        // address, so the public origin comes from configuration when set.
+        string baseUrl = string.IsNullOrWhiteSpace(publicBaseUrl)
+            ? $"{Request.Scheme}://{Request.Host}"
+            : publicBaseUrl.TrimEnd('/');
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply server/WebApi/Endpoints.cs]: Fixed in 1478a08: `Main` now takes `IHostEnvironment` and `ILogger<Main>`, and logs a warning (with the fallback scheme/host) when `Api:PublicBaseUrl` is unset outside `Development`. Added a test asserting the configured `PublicBaseUrl` wins over the request-derived host.

**Comment by DaveSkender** [comment]
Pushed 1478a08 addressing the open review feedback. Replied to each of the 18 inline threads (9 from CodeRabbit, 9 from the review swarm) with what changed and where.

A few items were only in the review bodies, not posted as separate inline threads, so noting disposition here:

- **`.vscode/extensions.json`** (CodeRabbit, outside diff): added `GitHub.copilot-chat` to the recommendations.
- **`server/WebApi.Tests/Endpoints/MainEndpointsTests.cs`** (CodeRabbit nitpick — test the `PublicBaseUrl` branch): added `GetIndicatorCatalog_WithPublicBaseUrlConfigured_UsesConfiguredOriginNotRequestHost`, asserting the configured origin wins over an internal request host.
- **`server/WebApi.Tests/Services/QuoteDatasetContractTests.cs`** (CodeRabbit nitpick — async fixture I/O): switched `File.ReadAllText` to `File.ReadAllTextAsync` with the test's cancellation token threaded through.
- **`web/src/config/env.ts`** (CodeRabbit nitpick + swarm finding — dedup `PROD_API_URL`): left as-is. `tests/vitepress/.vitepress/theme/index.ts` intentionally hardcodes the production URL and ignores `VITE_API_URL` in prod builds as a guard against a misconfigured Cloudflare Pages dashboard env var repointing the site at the wrong API. Both files already cross-reference each other in comments to catch drift at review time; centralizing the literal wouldn't remove the duplication without either keeping a second hardcoded fallback or dropping that guard. Open to revisiting if a shared constants file is preferred regardless.

Full validation run locally before pushing: `dotnet build`/`dotnet format --verify-no-changes`/`dotnet test` (65 passed), `pnpm run test:libs` (237 passed), web `tsc --noEmit && vite build`, web unit tests (77 passed), edge worker `tsc --noEmit` + `vitest run` (26 passed, including the two new spec files), `website-lint.sh check`, `lint:dotnet` (Roslynator clean), `lint:md`, and the VitePress package-consumption smoke test — all clean. CI is green on the new commit (build/test/lint/e2e/Codacy), with `Deploy` correctly gated behind the (skipped-on-preview) `deploy-api` job.

**Comment by DaveSkender** [comment]
Follow-up on 30f6281/02b4b5c: Codacy flagged one new issue in the container test (\`security/detect-object-injection\` on \`ApiContainer.outboundByHost?.[QUOTES_HOST]\`) — fixed by reading the handler via \`Object.entries\` instead of a computed-key lookup. Also added VS Code tasks for the two new CI gates (\`Lint/Test: Edge worker\`, \`Build: Validate container image (Docker)\`) so local runs match CI coverage.

All PR checks are green: build-backend, build-website, test-backend (65), test-libraries (237), test-web (77), test-e2e, Test edge worker (26, including the two new spec files), Validate container image, Codacy, and Deploy (correctly held on `deploy-api`'s result on main, skipped-but-passing on this preview).

File: .github/workflows/deploy-website.yml [hunk 19]
```diff
@@ -177,55 +160,88 @@ jobs:
...
       - build-website
+      - deploy-api
+    # deploy-api only runs on main; it is 'skipped' for preview builds, which
+    # must still be allowed to publish. On main, a failed API deploy must NOT
+    # be masked by a green Pages deployment.
+    if: |
+      always() &&
+      needs.build-backend.result == 'success' &&
+      needs.build-website.result == 'success' &&
+      (needs.deploy-api.result == 'success' || needs.deploy-api.result == 'skipped')
```
[the remark this answers is not in the document]
Comment (DaveSkender) [reply .github/workflows/deploy-website.yml:233]: Fixed in 074a054: added `needs.versioning.result == 'success'` to the `if:` condition alongside the existing checks, restoring the implicit all-`needs`-success gate that `always()` bypassed.

**Comment by DaveSkender** [comment]
## Production-corruption incident: root cause and fix

**Symptom:** every PR preview deploy was overwriting `charts.stockindicators.dev`.

**Root cause:** the `deploy` job in `deploy-website.yml` publishes to Cloudflare Pages via `wrangler pages deploy artifacts/web --project-name=stock-charts` with no `--branch`. That job has no `actions/checkout` step (it only downloads the built `web` artifact), so there's no `.git` directory for wrangler to read a branch from. Per Cloudflare's documented fallback, a deploy that can't resolve a branch is uploaded as a **production** deployment — confirmed directly from this PR's own CI logs (074a054's run): the output was a bare `https://<hash>.stock-charts.pages.dev` URL with no preview branch alias, which only happens on a production publish.

This predates the Cloudflare migration itself — `git log` on `deploy-website.yml` shows `--branch` was never passed, going back through every prior revision of this workflow. Every PR that touched `web/**` or `server/**` and ran this workflow has been overwriting production.

**Fix, in two parts (49968b7, c45c258):**
1. Pass `--branch=${{ github.head_ref || github.ref_name }}` explicitly, so only a push/dispatch on `main` is ever promoted to production.
2. First attempt routed the value through an `env:` var referenced as `"$DEPLOY_BRANCH"` inside `command:`, on the assumption `wrangler-action` runs that string through a shell. It doesn't — `command:` is executed via `@actions/exec`'s `ToolRunner`, which spawns wrangler directly with no shell, so `$DEPLOY_BRANCH` was never expanded (confirmed: that run published under the literal branch `deploy-branch`). Corrected to direct `${{ }}` interpolation, which is safe in this execution path since there's no shell to interpret metacharacters and git branch names can't contain the whitespace the tokenizer would need to split off extra arguments.

**Verified on this PR's head (c45c258):** the `Deploy` job now logs `--branch=migrate-to-cloudflare` and `Deployment alias URL: https://migrate-to-cloudflare.stock-charts.pages.dev` — a scoped preview, not production.

**Action needed post-merge:** because production has been serving whatever the *last* PR-preview build happened to contain (not necessarily `main`), please trigger a manual `workflow_dispatch` run of "Deploy website" from `main` after merging to republish known-good production content and confirm `charts.stockindicators.dev` is correct.
